pkg/visor/init.go

```diff
@@ -34,54 +35,132 @@ import (
+	// look at init function for module x. if init function makes use of
+	// field v.y for some module y of visor v, then y depends on x.
+	// if init function for x passes v itself, it's fucked up. All deps should be added, but
+	// better yet this should be refactored to pass the actual dep
+	up = vinit.MakeModule("updater", initUpdater, logger)
+	ebc = vinit.MakeModule("event_broadcaster", initEventBroadcaster, logger)
+	ar = vinit.MakeModule("address_resolver", initAddressResolver, logger)
+	disc = vinit.MakeModule("discovery", initDiscovery, logger)
+	sn = vinit.MakeModule("snet", initSNet, logger, &ar, &disc, &ebc)
+	pty = vinit.MakeModule("dmsg_pty", initDmsgpty, logger, &sn)
+	tr = vinit.MakeModule("transport", initTransport, logger, &sn, &ebc)
+	rt = vinit.MakeModule("router", initRouter, logger, &tr, &sn)
+	launch = vinit.MakeModule("launcher", initLauncher, logger, &ebc, &disc, &sn, &tr, &rt)
+	cli = vinit.MakeModule("cli", initCLI, logger)
+	hvs = vinit.MakeModule("hypervisors", initHypervisors, logger, &sn)
+	ut = vinit.MakeModule("uptime_tracker", initUptimeTracker, logger)
+	trv = vinit.MakeModule("trusted_visors", initTrustedVisors, logger, &tr)
+	vis = vinit.MakeModule("visor", vinit.DoNothing, logger, &up, &ebc, &ar, &disc, &sn, &pty,
+		&tr, &rt, &launch, &cli, &hvs, &ut, &trv)
+	hv = vinit.MakeModule("hypervisor", initHypervisor, logger, &vis)
 }
 
-func initUpdater(v *Visor) bool {
-	report := v.makeReporter("updater")
+// -------------------------------------------------------------------
 
+func initUpdater(ctx context.Context, log *logging.Logger) error {
+	v, err := getVisor(ctx)
+	if err != nil {
+		return err
+	}
+	updater := updater.New(v.log, v.restartCtx, v.conf.Launcher.BinPath)
+
+	v.initLock.Lock()
+	defer v.initLock.Unlock()
 	v.restartCtx.SetCheckDelay(time.Duration(v.conf.RestartCheckDelay))
 	v.restartCtx.RegisterLogger(v.log)
-	v.updater = updater.New(v.log, v.restartCtx, v.conf.Launcher.BinPath)
-	return report(nil)
+	v.updater = updater
+	return nil
 }
 
-func initEventBroadcaster(v *Visor) bool {
+func initEventBroadcaster(ctx context.Context, log *logging.Logger) error {
+	v, err := getVisor(ctx)
+	if err != nil {
+		return err
+	}
 	report := v.makeReporter("event_broadcaster")
 
-	log := v.MasterLogger().PackageLogger("event_broadcaster")
 	const ebcTimeout = time.Second
 	ebc := appevent.NewBroadcaster(log, ebcTimeout)
-
 	v.pushCloseStack("event_broadcaster", func() bool {
 		return report(ebc.Close())
 	})
 
+	v.initLock.Lock()
 	v.ebc = ebc
-	return report(nil)
+	v.initLock.Unlock()
+	return nil
+}
+
+func initAddressResolver(ctx context.Context, log *logging.Logger) error {
+	v, err := getVisor(ctx)
+	if err != nil {
+		return err
+	}
+	report := v.makeReporter("address-resolver")
+	conf := v.conf.Transport
+
+	arClient, err := arclient.NewHTTP(conf.AddressResolver, v.conf.PK, v.conf.SK)
+	if err != nil {
+		err := fmt.Errorf("failed to create address resolver client: %w", err)
+		report(err)
```

Comment:
Yes, it is weird. I'm strongly leaning towards removing report and corresponding calls to it. Basically, I left it untouched for the time being. Currently it is used to check for errors in concurrent code, but in my opinion it's faulty and can skip errors. So that's definitely an area to improve yet